srml/contract/src/lib.rs

```diff
@@ -92,6 +93,50 @@ pub trait ComputeDispatchFee<Call, Balance> {
 	fn compute_dispatch_fee(call: &Call) -> Balance;
 }
 
+#[derive(Encode,Decode,Clone,Debug)]
+/// Information for managing a sub trie abstraction.
+/// This is the required info to cache for an account
+pub struct SubTrie {
+	/// unique ID for the subtree encoded as a byte
+	pub key_space: KeySpace,
+	/// the size of stored value in octet
+	pub current_mem_stored: u64,
+}
+
+/// Get a key space (key space must be unique and collision resistant depending upon its context)
+/// Note that it is different than encode because key space should have collision resistance
+/// property (being a proper uniqueid).
+pub trait KeySpaceGenerator<AccountId: Sized> {
+	/// get a keyspace for an account, using reference to parent account keyspace to ensure
+	/// uniqueness of keyspace
+	/// The implementation must ensure every new key space is unique: two consecutive call with the
+	/// same parameter needs to return different key space values.
+	fn key_space(account_id: &AccountId) -> KeySpace;
+}
+
+/// Get key space from `account_id`
+pub struct KeySpaceFromParentCounter<T: Trait>(PhantomData<T>);
+
+/// This generator use inner counter for account id and apply hash over `AccountId +
+/// parent_keyspace + accountid_counter`
+impl<T: Trait> KeySpaceGenerator<T::AccountId> for KeySpaceFromParentCounter<T>
+where
+	T::AccountId: AsRef<[u8]> {
+	fn key_space(account_id: &T::AccountId) -> KeySpace {
+		let new_seed = <AccountCounter<T>>::get().wrapping_add(1);
+
+		let mut buf = Vec::new();
+		buf.extend_from_slice(account_id.as_ref());
+		buf.extend_from_slice(&new_seed.to_be_bytes()[..]);
+		// this line is yelling use Hash instead of vec: would require making KeySpace a subtype of
+		// Trait
+		let res = T::Hashing::hash(&buf[..]).as_ref().into();
+	
+		<AccountCounter<T>>::put(new_seed);
```

Comment:
I like committed only on the end of the method since there is no rollback in case of error (no reason for error at this point). But it is true that worst case scenario we are skipping a value that does not need to be sequential. So I can use that (second TODO after decl_storage).